webapp/login.go

```diff
@@ -0,0 +1,323 @@
+	var err error
+	value := map[string]interface{}{
+		"user":  *user,
+		"token": *token,
+	}
+
+	sc, err := getSecureCookie(ctx)
+	if err != nil {
+		return err
+	}
+
+	if encoded, err := sc.Encode("session", value); err == nil {
+		cookie := &http.Cookie{
+			Name:     "session",
+			Value:    encoded,
+			Path:     "/",
+			MaxAge:   2592000,
+			HttpOnly: true,
+			Secure:   true,
+		}
+
+		// SameSite=None for http.Cookie is only available in Go.113;
+		// see https://github.com/golang/go/issues/32546.
+		if v := cookie.String(); v != "" {
+			response.Header().Add("Set-Cookie", v+"; SameSite=None")
+		}
+	} else {
+		log := shared.GetLogger(ctx)
+		log.Errorf("Failed to set session cookie: %s", err.Error())
+	}
+
+	return err
+}
+
+func setState(ctx context.Context, state string, response http.ResponseWriter) error {
+	var err error
+	sc, err := getSecureCookie(ctx)
+	if err != nil {
+		return err
+	}
+
+	if encoded, err := sc.Encode("state", state); err == nil {
+		cookie := &http.Cookie{
+			Name:     "state",
+			Value:    encoded,
+			Path:     "/",
+			MaxAge:   600,
+			Secure:   true,
+			HttpOnly: true,
+			SameSite: http.SameSiteLaxMode,
+		}
+		http.SetCookie(response, cookie)
+	}
+
+	return err
+}
+
+func getState(r *http.Request) string {
+	ctx := shared.NewAppEngineContext(r)
+	log := shared.GetLogger(ctx)
+	cookieValue := ""
+	if cookie, err := r.Cookie("state"); err == nil && cookie != nil {
+		sc, err := getSecureCookie(ctx)
+		if err != nil {
+			return ""
+		}
+
+		if err = sc.Decode("state", cookie.Value, &cookieValue); err != nil {
+			log.Errorf("Failed to decode cookie for state: %s", err.Error())
+		}
+	} else {
+		log.Errorf("Failed to get state cookie: %s", err.Error())
+	}
+	return cookieValue
+}
+
+var secureCookie *securecookie.SecureCookie
+
+func getSecureCookie(ctx context.Context) (*securecookie.SecureCookie, error) {
+	log := shared.GetLogger(ctx)
+	if secureCookie == nil {
+		store := shared.NewAppEngineDatastore(ctx, false)
+		hashKey, err := shared.GetSecret(store, "secure-cookie-hashkey")
+		if err != nil {
+			log.Errorf("Failed to get secure-cookie-hashkey secret: %s", err.Error())
+			return nil, err
+		}
+
+		blockKey, err := shared.GetSecret(store, "secure-cookie-blockkey")
+		if err != nil {
+			log.Errorf("Failed to get secure-cookie-blockkey secret: %s", err.Error())
+			return nil, err
+		}
+
+		secureCookie = securecookie.New([]byte(hashKey), []byte(blockKey))
+	}
+	return secureCookie, nil
+}
+
+func clearSession(response http.ResponseWriter) {
+	cookie := &http.Cookie{
+		Name:   "session",
+		Value:  "",
+		Path:   "/",
+		MaxAge: -1,
+	}
+	http.SetCookie(response, cookie)
+}
+
+func getGithubOAuthConfig(ctx context.Context) *oauth2.Config {
+	store := shared.NewAppEngineDatastore(ctx, false)
+	log := shared.GetLogger(ctx)
+	clientID, err := shared.GetSecret(store, "github-oauth-client-id")
+	if err != nil {
+		log.Errorf("Failed to get github-oauth-client-id secret: %s", err.Error())
```

Comment:
When this happens, the returned `Config` will be unusable (without a `clientID`), but the caller couldn't tell unless they inspect the members. Since this is a fatal error, we generally want to return the error, too, so that the caller can act accordingly (fail early).